src/server/pkg/storage/fileset/reader.go

```diff
@@ -1,43 +1,56 @@
 package fileset
 
 import (
-	"bytes"
 	"context"
 	"io"
 
 	"github.com/gogo/protobuf/proto"
-	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
 	"github.com/pachyderm/pachyderm/src/server/pkg/storage/chunk"
 	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
-	"github.com/pachyderm/pachyderm/src/server/pkg/tar"
 )
 
 // Reader is an abstraction for reading a fileset.
 type Reader struct {
-	chunks       *chunk.Storage
-	indexFactory func(ctx context.Context) (*index.Reader, error)
+	store              Store
+	chunks             *chunk.Storage
+	path               string
+	indexOpts          []index.Option
+	additive, deletive *index.Reader
 }
 
 func newReader(store Store, chunks *chunk.Storage, p string, opts ...index.Option) *Reader {
-	return &Reader{
-		chunks: chunks,
-		indexFactory: func(ctx context.Context) (*index.Reader, error) {
-			topIdx, err := store.GetIndex(ctx, p)
-			if err != nil {
-				return nil, err
-			}
-			return index.NewReader(chunks, topIdx, opts...), nil
-		},
+	r := &Reader{
+		store:     store,
+		chunks:    chunks,
+		path:      p,
+		indexOpts: opts,
 	}
+	return r
 }
 
-// Iterate iterates over the files in the fileset.
-func (r *Reader) Iterate(ctx context.Context, cb func(File) error) error {
-	ir, err := r.indexFactory(ctx)
-	if err != nil {
+func (r *Reader) setup(ctx context.Context) error {
+	if r.additive == nil {
+		md, err := r.store.Get(ctx, r.path)
+		if err != nil {
+			return err
+		}
+		r.additive = index.NewReader(r.chunks, md.Additive, r.indexOpts...)
```

Comment:
Can we return these? `setup(ctx) (additive, deletive, error)` and remove the associated fields from the reader?

Reply:
Yeah, that is a good idea.